content/post/2019-05-18-talking-i2c.md

```diff
@@ -0,0 +1,145 @@
+
+## Analog to Digital
+
+Analog to digital converters take in an analog signal like the `0-3` volt signal described above, and turn it into a digital signal.
+I'm using an [ADS1115](http://www.ti.com/product/ADS1115) chip which uses an [i²c interface](https://i2c.info/) so that you can read 4 different analog signals with just one pair of wires.
+This allows you to read each of the analog signals as a 2-byte value, a number between `-32,768` and `32,767`.
+The raspberry pi has built-in support for talking to I2C devices, so this is a great option for my use-case.
+
+## Eye to See
+
+The I2C protocol allows you to have multiple devices connected to the same pair of wires that act as a message bus.
+The raspberry pi acts as a master, meaning that it initiates communication.
+Each device on the bus has an address, and you always use that address when sending or receiving data.
+
+Because the protocol is so basic, many devices implement a [register-based protocol](http://www.ti.com/lit/an/slva704/slva704.pdf).
+Each time you want to send a message to the device, you send a first byte that identifies which register you are writing to and the rest of the bytes are the information you want to write into that register.
+If you want to ask for some information you write 1 byte to identify what you are asking about, and then you read a certain number of bytes from that address to get the data you want.
+
+## Checking the Soil
+
+The first thing we'll do is check the configuration of the device.
+To do this well send 1 byte telling the chip which register we want to sent and then read back 2 bytes (the size of the configuration register).
+
+```elixir
+{:ok, bus} = Circuits.I2C.open("i2c-1")
+:ok = Circuits.I2C.write(bus, 72, <<1>>)
+{:ok, <<133, 131>>} = Circuits.I2C.read(bus, 72, 2)
+# I2C also has a convenience method for this
+{:ok, <<133, 131>>} = Circuits.I2C.write_read(bus, 72, <<1>>, 2)
+```
+
+We're getting back some data, but what do those bytes mean?
+We'll need to take a look at [the datasheet for the ADS1115](http://www.ti.com/lit/ds/symlink/ads1115.pdf) chip.
+These 16 bits of data tell us 9 different things about how the chip is currently operating.
+That's a lot of information packed into the space of just two ascii characters.
+
+![Config Register](/images/2019-05-18/ADS1115_Config_Register.png)
+
+Luckily, we can use [Elixir's binary pattern matching syntax](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#%3C%3C%3E%3E/1) to pull out this information.
+Binary pattern matching allows us to ask for information by expressing how we expect the information to be laid out across the bytes.
+No bit-shifting required.
+
+```elixir
+<< status::size(1), mux::size(3), _other_stuff::size(12) >> = <<133, 131>>
+```
+
+We can also similarly build a new binary to write new configuration to the register.
+
+```elixir
+config_bytes = << status::size(1), mux::size(3), other_stuff::size(12) >>
+Circuits.I2C.write(bus, 72, <<1>> <> config_bytes)
```

Comment:
I think that the binary concatenation is pretty understandable here, so don't change. However, you can also write iodata, `Circuits.I2C.write(bus, 72, [1, config_bytes])`, or use binary syntax, `<<1, config_bytes::binary>>`.  If you were going to show more code that needed it, it would be nice to change, but since this is so simple, it doesn't seem worthwhile to have to explain iodata.

Reply:
I was wondering whether `iodata` would be supported, but had not tried it yet. I agree that explaining `iodata` feels a little out-of-scope for this post